### ISSUE TITLE
perl5db.pl: Undefined subroutine &DB::db_warn called

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14400]
Migrated from [rt.perl.org#123553](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123553) (status was 'resolved')

Searchable as RT123553$

**Comment by p5pRT** [comment]
### From @choroba

#### Created by @choroba



Perl debugger sometimes tries to call a non\-existent subroutine
db\_warn\. It's real name is \_db\_warn\, though\.

How to replicate&#8203;:

perl \-d \-e1
\!/

Output&#8203;:

/bin/bash&#8203;: /&#8203;: Is a directory
Undefined subroutine &DB&#8203;::db\_warn called at /usr/lib/perl5/5\.18\.1/perl5db\.pl line 6740\.
&nbsp;  at /usr/lib/perl5/5\.18\.1/perl5db\.pl line 6740\.
&nbsp;         DB&#8203;::\_db\_system\('/bin/bash'\, '\-c'\, '/'\) called at /usr/lib/perl5/5\.18\.1/perl5db\.pl line 3923
&nbsp;         DB&#8203;::Obj&#8203;::\_handle\_sh\_command\('DB&#8203;::Obj=HASH\(0x105df48\)'\) called at /usr/lib/perl5/5\.18\.1/perl5db\.pl line 2992
&nbsp;         DB&#8203;::DB called at \-e line 1

Patch to fix the issue attached\.

&nbsp;         Choroba

<details><summary>Perl Info</summary>

```

Flags:
     category=library
     severity=low
     module=DB

This perlbug was built using Perl 5.18.1 - Fri Sep 27 18:27:33 UTC 2013
It is being executed now by  Perl 5.18.1 - Fri Sep 27 18:20:30 UTC 2013.

Site configuration information for perl 5.18.1:

Configured by abuild at Fri Sep 27 18:20:30 UTC 2013.

Summary of my perl5 (revision 5 version 18 subversion 1) configuration:

   Platform:
     osname=linux, osvers=3.4.6-2.10-xen, archname=x86_64-linux-thread-multi
     uname='linux build16 3.4.6-2.10-xen #1 smp thu jul 26 09:36:26 utc 2012 
(641c197) x86_64 x86_64 x86_64 gnulinux '
     config_args='-ds -e -Dprefix=/usr -Dvendorprefix=/usr -Dinstallusrbinperl 
-Dusethreads -Di_db -Di_dbm -Di_ndbm -Di_gdbm -Dd_dbm_open -Duseshrplib=true 
-Doptimize=-fmessage-length=0 -grecord-gcc-switches -O2 -Wall 
-D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables 
-fasynchronous-unwind-tables -g -Wall -pipe -Accflags=-DPERL_USE_SAFE_PUTENV 
-Dotherlibdirs=/usr/lib/perl5/site_perl 
-Dinc_version_list=5.18.0/x86_64-linux-thread-multi 5.18.0'
     hint=recommended, useposix=true, d_sigaction=define
     useithreads=define, usemultiplicity=define
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=define, use64bitall=define, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DPERL_USE_SAFE_PUTENV 
-fno-strict-aliasing -pipe -fstack-protector -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64',
     optimize='-fmessage-length=0 -grecord-gcc-switches -O2 -Wall 
-D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables 
-fasynchronous-unwind-tables -g -Wall -pipe',
     cppflags='-D_REENTRANT -D_GNU_SOURCE -DPERL_USE_SAFE_PUTENV 
-fno-strict-aliasing -pipe -fstack-protector'
     ccversion='', gccversion='4.8.1 20130909 [gcc-4_8-branch revision 
202388]', gccosandvers=''
     intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
     ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='cc', ldflags =' -L/usr/local/lib64 -fstack-protector'
     libpth=/lib64 /usr/lib64 /usr/local/lib64
     libs=-lm -ldl -lcrypt -lpthread
     perllibs=-lm -ldl -lcrypt -lpthread
     libc=/lib64/libc-2.18.so, so=so, useshrplib=true, libperl=libperl.so
     gnulibc_version='2.18'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E 
-Wl,-rpath,/usr/lib/perl5/5.18.1/x86_64-linux-thread-multi/CORE'
     cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib64 
-fstack-protector'

Locally applied patches:



@INC for perl 5.18.1:
     /home/choroba/perl5/lib/perl5/x86_64-linux-thread-multi
     /home/choroba/perl5/lib/perl5/x86_64-linux-thread-multi
     /home/choroba/perl5/lib/perl5
     /home/choroba/perl5/lib/perl5/x86_64-linux-thread-multi
     /home/choroba/perl5/lib/perl5/x86_64-linux-thread-multi
     /home/choroba/perl5/lib/perl5
     /usr/lib/perl5/site_perl/5.18.1/x86_64-linux-thread-multi
     /usr/lib/perl5/site_perl/5.18.1
     /usr/lib/perl5/vendor_perl/5.18.1/x86_64-linux-thread-multi
     /usr/lib/perl5/vendor_perl/5.18.1
     /usr/lib/perl5/5.18.1/x86_64-linux-thread-multi
     /usr/lib/perl5/5.18.1
     /usr/lib/perl5/site_perl
     .


Environment for perl 5.18.1:
     HOME=/home/choroba
     LANG=en_US.UTF-8
     LANGUAGE=en_US.utf8
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)

PATH=/home/choroba/perl5/bin:/home/choroba/bin:/bin:/sbin:/usr/bin:/usr/local/bin:/usr/games:/usr/X11R6/bin:/opt/gnome/bin:/home/choroba/perl/bin:.

PERL5LIB=/home/choroba/perl5/lib/perl5/x86_64-linux-thread-multi:/home/choroba/perl5/lib/perl5:/home/choroba/perl5/lib/perl5/x86_64-linux-thread-multi:/home/choroba/perl5/lib/perl5:
     PERL_BADLANG (unset)
     PERL_LOCAL_LIB_ROOT=/home/choroba/perl5
     PERL_MB_OPT=--install_base /home/choroba/perl5
     PERL_MM_OPT=INSTALL_BASE=/home/choroba/perl5
     SHELL=/bin/bash
```
</details>

**Comment by p5pRT** [comment]
### From @choroba

<details><summary>perl5db.patch</summary>

```diff
--- perl5db.pl.backup	2015-01-06 00:35:29.089785377 +0100
+++ perl5db.pl	2015-01-06 00:35:56.181979984 +0100
@@ -6723,24 +6723,24 @@
 
     # We save, change, then restore STDIN and STDOUT to avoid fork() since
     # some non-Unix systems can do system() but have problems with fork().
-    open( SAVEIN,  "<&STDIN" )  || db_warn("Can't save STDIN");
-    open( SAVEOUT, ">&STDOUT" ) || db_warn("Can't save STDOUT");
-    open( STDIN,   "<&IN" )     || db_warn("Can't redirect STDIN");
-    open( STDOUT,  ">&OUT" )    || db_warn("Can't redirect STDOUT");
+    open( SAVEIN,  "<&STDIN" )  || _db_warn("Can't save STDIN");
+    open( SAVEOUT, ">&STDOUT" ) || _db_warn("Can't save STDOUT");
+    open( STDIN,   "<&IN" )     || _db_warn("Can't redirect STDIN");
+    open( STDOUT,  ">&OUT" )    || _db_warn("Can't redirect STDOUT");
 
     # XXX: using csh or tcsh destroys sigint retvals!
     system(@_);
-    open( STDIN,  "<&SAVEIN" )  || db_warn("Can't restore STDIN");
-    open( STDOUT, ">&SAVEOUT" ) || db_warn("Can't restore STDOUT");
+    open( STDIN,  "<&SAVEIN" )  || _db_warn("Can't restore STDIN");
+    open( STDOUT, ">&SAVEOUT" ) || _db_warn("Can't restore STDOUT");
     close(SAVEIN);
     close(SAVEOUT);
 
     # most of the $? crud was coping with broken cshisms
     if ( $? >> 8 ) {
-        db_warn( "(Command exited ", ( $? >> 8 ), ")\n" );
+        _db_warn( "(Command exited ", ( $? >> 8 ), ")\n" );
     }
     elsif ($?) {
-        db_warn(
+        _db_warn(
             "(Command died of SIG#",
             ( $? & 127 ),
             ( ( $? & 128 ) ? " -- core dumped" : "" ),

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Jan 05 16&#8203;:26&#8203;:08 2015\, choroba@&#8203;cpan\.org wrote&#8203;:
> This is a bug report for perl from choroba@&#8203;cpan\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.18\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> Perl debugger sometimes tries to call a non\-existent subroutine
> db\_warn\. It's real name is \_db\_warn\, though\.
> 
> How to replicate&#8203;:
> 
> perl \-d \-e1
> \!/
> 
> Output&#8203;:
> 
> /bin/bash&#8203;: /&#8203;: Is a directory
> Undefined subroutine &DB&#8203;::db\_warn called at
> /usr/lib/perl5/5\.18\.1/perl5db\.pl line 6740\.
>   at /usr/lib/perl5/5\.18\.1/perl5db\.pl line 6740\.
>          DB&#8203;::\_db\_system\('/bin/bash'\, '\-c'\, '/'\) called at
> /usr/lib/perl5/5\.18\.1/perl5db\.pl line 3923
>          DB&#8203;::Obj&#8203;::\_handle\_sh\_command\('DB&#8203;::Obj=HASH\(0x105df48\)'\) called
> at /usr/lib/perl5/5\.18\.1/perl5db\.pl line 2992
>          DB&#8203;::DB called at \-e line 1
> 
> Patch to fix the issue attached\.
> 
> Choroba
> 


Thank you very much for this report\.  This is a bug which crept in somewhere between perl 5\.16\.3 and perl 5\.18\.2\.

\#\#\#\#\#
\[~\] 11 $ perlbrew switch perl\-5\.16\.3
\[~\] 12 $ perl \-d \-e1

Loading DB routines from perl5db\.pl version 1\.37
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    1
&nbsp;  DB\<1> \!/
No such command\!

&nbsp;  DB\<1> q
\[~\] 13 $ perlbrew switch perl\-5\.18\.2
\[~\] 14 $ perl \-d \-e1

Loading DB routines from perl5db\.pl version 1\.39\_10
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    1
&nbsp;  DB\<1> \!/
/bin/bash&#8203;: /&#8203;: Is a directory
Undefined subroutine &DB&#8203;::db\_warn called at /home/jkeenan/perl5/perlbrew/perls/perl\-5\.18\.2/lib/5\.18\.2/perl5db\.pl line 6735\.
 at /home/jkeenan/perl5/perlbrew/perls/perl\-5\.18\.2/lib/5\.18\.2/perl5db\.pl line 6735\.
&nbsp;    DB&#8203;::\_db\_system\('/bin/bash'\, '\-c'\, '/'\) called at /home/jkeenan/perl5/perlbrew/perls/perl\-5\.18\.2/lib/5\.18\.2/perl5db\.pl line 3918
&nbsp;    DB&#8203;::Obj&#8203;::\_handle\_sh\_command\('DB&#8203;::Obj=HASH\(0x1dee6b0\)'\) called at /home/jkeenan/perl5/perlbrew/perls/perl\-5\.18\.2/lib/5\.18\.2/perl5db\.pl line 2987
&nbsp;    DB&#8203;::DB called at \-e line 1
Debugged program terminated\.  Use q to quit or R to restart\,
use o inhibit\_exit to avoid stopping after program termination\,
h q\, h R or h o to get additional info\.
&nbsp;  DB\<2> q
\#\#\#\#\#

We should figure out what commit broke this functionality\, then consider whether we should apply your patch or try to restore the previous behavior\.


> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>      category=library
>      severity=low
>      module=DB
> \-\-\-
> This perlbug was built using Perl 5\.18\.1 \- Fri Sep 27 18&#8203;:27&#8203;:33 UTC
> 2013
> It is being executed now by  Perl 5\.18\.1 \- Fri Sep 27 18&#8203;:20&#8203;:30 UTC
> 2013\.
> 
> Site configuration information for perl 5\.18\.1&#8203;:
> 
> Configured by abuild at Fri Sep 27 18&#8203;:20&#8203;:30 UTC 2013\.
> 
> Summary of my perl5 \(revision 5 version 18 subversion 1\)
> configuration&#8203;:
> 
> Platform&#8203;:
>   osname=linux\, osvers=3\.4\.6\-2\.10\-xen\, archname=x86\_64\-linux\-thread\-
> multi
>    uname='linux build16 3\.4\.6\-2\.10\-xen \#1 smp thu jul 26 09&#8203;:36&#8203;:26 utc
> 2012
> \(641c197\) x86\_64 x86\_64 x86\_64 gnulinux '
>    config\_args='\-ds \-e \-Dprefix=/usr \-Dvendorprefix=/usr
> \-Dinstallusrbinperl
> \-Dusethreads \-Di\_db \-Di\_dbm \-Di\_ndbm \-Di\_gdbm \-Dd\_dbm\_open
> \-Duseshrplib=true
> \-Doptimize=\-fmessage\-length=0 \-grecord\-gcc\-switches \-O2 \-Wall
> \-D\_FORTIFY\_SOURCE=2 \-fstack\-protector \-funwind\-tables
> \-fasynchronous\-unwind\-tables \-g \-Wall \-pipe \-Accflags=\-
> DPERL\_USE\_SAFE\_PUTENV
> \-Dotherlibdirs=/usr/lib/perl5/site\_perl
> \-Dinc\_version\_list=5\.18\.0/x86\_64\-linux\-thread\-multi 5\.18\.0'
>   hint=recommended\, useposix=true\, d\_sigaction=define
>   useithreads=define\, usemultiplicity=define
>   useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>   use64bitint=define\, use64bitall=define\, uselongdouble=undef
>   usemymalloc=n\, bincompat5005=undef
> Compiler&#8203;:
>    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE
> \-DPERL\_USE\_SAFE\_PUTENV
> \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-D\_LARGEFILE\_SOURCE
> \-D\_FILE\_OFFSET\_BITS=64'\,
>    optimize='\-fmessage\-length=0 \-grecord\-gcc\-switches \-O2 \-Wall
> \-D\_FORTIFY\_SOURCE=2 \-fstack\-protector \-funwind\-tables
> \-fasynchronous\-unwind\-tables \-g \-Wall \-pipe'\,
>    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DPERL\_USE\_SAFE\_PUTENV
> \-fno\-strict\-aliasing \-pipe \-fstack\-protector'
>    ccversion=''\, gccversion='4\.8\.1 20130909 \[gcc\-4\_8\-branch revision
> 202388\]'\, gccosandvers=''
>   intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
>   d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
> lseeksize=8
>   alignbytes=8\, prototype=define
> Linker and Libraries&#8203;:
>   ld='cc'\, ldflags =' \-L/usr/local/lib64 \-fstack\-protector'
>   libpth=/lib64 /usr/lib64 /usr/local/lib64
>   libs=\-lm \-ldl \-lcrypt \-lpthread
>   perllibs=\-lm \-ldl \-lcrypt \-lpthread
>   libc=/lib64/libc\-2\.18\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
>   gnulibc\_version='2\.18'
> Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E
> \-Wl\,\-rpath\,/usr/lib/perl5/5\.18\.1/x86\_64\-linux\-thread\-multi/CORE'
>    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib64
> \-fstack\-protector'
> 
> Locally applied patches&#8203;:
> 
> 
> \-\-\-
> @&#8203;INC for perl 5\.18\.1&#8203;:
>      /home/choroba/perl5/lib/perl5/x86\_64\-linux\-thread\-multi
>      /home/choroba/perl5/lib/perl5/x86\_64\-linux\-thread\-multi
>      /home/choroba/perl5/lib/perl5
>      /home/choroba/perl5/lib/perl5/x86\_64\-linux\-thread\-multi
>      /home/choroba/perl5/lib/perl5/x86\_64\-linux\-thread\-multi
>      /home/choroba/perl5/lib/perl5
>      /usr/lib/perl5/site\_perl/5\.18\.1/x86\_64\-linux\-thread\-multi
>      /usr/lib/perl5/site\_perl/5\.18\.1
>      /usr/lib/perl5/vendor\_perl/5\.18\.1/x86\_64\-linux\-thread\-multi
>      /usr/lib/perl5/vendor\_perl/5\.18\.1
>      /usr/lib/perl5/5\.18\.1/x86\_64\-linux\-thread\-multi
>      /usr/lib/perl5/5\.18\.1
>      /usr/lib/perl5/site\_perl
>      \.
> 
> \-\-\-
> Environment for perl 5\.18\.1&#8203;:
>      HOME=/home/choroba
>      LANG=en\_US\.UTF\-8
>      LANGUAGE=en\_US\.utf8
>      LD\_LIBRARY\_PATH \(unset\)
>      LOGDIR \(unset\)
> 
> PATH=/home/choroba/perl5/bin&#8203;:/home/choroba/bin&#8203;:/bin&#8203;:/sbin&#8203;:/usr/bin&#8203;:/usr/local/bin&#8203;:/usr/games&#8203;:/usr/X11R6/bin&#8203;:/opt/gnome/bin&#8203;:/home/choroba/perl/bin&#8203;:\.
> 
> PERL5LIB=/home/choroba/perl5/lib/perl5/x86\_64\-linux\-thread\-
> multi&#8203;:/home/choroba/perl5/lib/perl5&#8203;:/home/choroba/perl5/lib/perl5/x86\_64\-
> linux\-thread\-multi&#8203;:/home/choroba/perl5/lib/perl5&#8203;:
>      PERL\_BADLANG \(unset\)
>      PERL\_LOCAL\_LIB\_ROOT=/home/choroba/perl5
>      PERL\_MB\_OPT=\-\-install\_base /home/choroba/perl5
>      PERL\_MM\_OPT=INSTALL\_BASE=/home/choroba/perl5
>      SHELL=/bin/bash


\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Jan 05 18&#8203;:51&#8203;:14 2015\, jkeenan wrote&#8203;:
> We should figure out what commit broke this functionality\, then
> consider whether we should apply your patch or try to restore the
> previous behavior\.

b5679dc0b0 clearly had a few typos in it\.  That 5\.16 had no \! command is unrelated\. 

On Mon Jan 05 16&#8203;:26&#8203;:08 2015\, choroba@&#8203;cpan\.org wrote&#8203;:
> Patch to fix the issue attached\.

Thank you\.  Applied as 2384afee950\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
